File: src/trading/drift.ts
import type { PlanetMarketState } from "../data/types/planetMarketState.ts";
import type { TradeDirection } from "../data/types/tradeDirection.ts";
import { clamp } from "../simulation/clamp.ts";
import {
  BASELINE_DRIFT_PERCENT,
  PRICE_FLOOR_PERCENT,
  PRICE_CEILING_PERCENT,
  PRICE_RECOVERY_PERCENT_PER_HOUR,
} from "../data/constants/tradingConfig.ts";

function bounds(basePrice: number): { floor: number; ceiling: number } {
  return { floor: basePrice * PRICE_FLOOR_PERCENT, ceiling: basePrice * PRICE_CEILING_PERCENT };
}

// Phase 3 GDD §2.6. Direction is the trade's direction from the market's
// perspective: 'sell' = a player sold INTO this market (added supply),
// dropping the price; 'buy' = a player bought FROM this market (removed
// supply), raising it. Applied one unit at a time (not as a single
// compounded exponent) so the floor/ceiling clamp holds even transiently
// mid-calculation, per the contract's explicit requirement.
export function applyDrift(
  marketState: PlanetMarketState,
  unitsTraded: number,
  direction: TradeDirection,
): PlanetMarketState {
  const { floor, ceiling } = bounds(marketState.basePrice);
  const sign = direction === "buy" ? 1 : -1;

  let price = marketState.currentPrice;
  for (let i = 0; i < unitsTraded; i++) {
    price = clamp(price * (1 + sign * BASELINE_DRIFT_PERCENT), floor, ceiling);
  }

  return { ...marketState, currentPrice: price };
}

// Phase 3 GDD §2.6 -- drifts currentPrice back toward basePrice over time
// when untraded. Exponential decay of the remaining gap (percentage of the
// gap closed per hour), consistent with the drift formula's own
// percentage-based/naturally-diminishing design rather than a linear walk
// that could overshoot basePrice. elapsedHours may be fractional.
export function applyRecovery(marketState: PlanetMarketState, elapsedHours: number): PlanetMarketState {
  const gap = marketState.basePrice - marketState.currentPrice;
  const remainingGap = gap * (1 - PRICE_RECOVERY_PERCENT_PER_HOUR) ** elapsedHours;
  const price = marketState.basePrice - remainingGap;

  return { ...marketState, currentPrice: price };
}

File: unity/ProfitableCore/Simulation/Drift.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/trading/drift.ts.
public static class Drift
{
    private static (double Floor, double Ceiling) Bounds(double basePrice) =>
        (basePrice * TradingConfig.PriceFloorPercent, basePrice * TradingConfig.PriceCeilingPercent);

    // Direction is the trade's direction from the market's perspective:
    // Sell = a player sold INTO this market (added supply), dropping the
    // price; Buy = a player bought FROM this market (removed supply),
    // raising it. Applied one unit at a time (not as a single compounded
    // exponent) so the floor/ceiling clamp holds even transiently
    // mid-calculation, per the source's own explicit requirement.
    public static PlanetMarketState ApplyDrift(PlanetMarketState marketState, int unitsTraded, TradeDirection direction)
    {
        var (floor, ceiling) = Bounds(marketState.BasePrice);
        var sign = direction == TradeDirection.Buy ? 1 : -1;

        var price = marketState.CurrentPrice;
        for (var i = 0; i < unitsTraded; i++)
        {
            price = ClampHelper.Clamp(price * (1 + sign * TradingConfig.BaselineDriftPercent), floor, ceiling);
        }

        return new PlanetMarketState
        {
            PlanetId = marketState.PlanetId,
            ItemId = marketState.ItemId,
            BasePrice = marketState.BasePrice,
            CurrentPrice = price,
        };
    }

    // Drifts CurrentPrice back toward BasePrice over time when untraded.
    // Exponential decay of the remaining gap (percentage of the gap
    // closed per hour), consistent with the drift formula's own
    // percentage-based/naturally-diminishing design rather than a linear
    // walk that could overshoot BasePrice. elapsedHours may be fractional.
    public static PlanetMarketState ApplyRecovery(PlanetMarketState marketState, double elapsedHours)
    {
        var gap = marketState.BasePrice - marketState.CurrentPrice;
        var remainingGap = gap * Math.Pow(1 - TradingConfig.PriceRecoveryPercentPerHour, elapsedHours);
        var price = marketState.BasePrice - remainingGap;

        return new PlanetMarketState
        {
            PlanetId = marketState.PlanetId,
            ItemId = marketState.ItemId,
            BasePrice = marketState.BasePrice,
            CurrentPrice = price,
        };
    }
}
